lib/ansible/modules/cloud/vmware/vcenter_folder.py

```diff
@@ -0,0 +1,253 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: vcenter_folder
+short_description: Add/remove folders to/from vCenter
+description:
+- This module can be used to add/remove a folder to/from vCenter.
+version_added: '2.4'
+author:
+- Davis Phillips (@dav1x)
+- Dag Wieers (@dagwieers)
+notes:
+- Tested on vSphere 6.5
+requirements:
+- PyVmomi
+options:
+  datacenter:
+    description:
+    - Name of the datacenter to add the host.
+    required: yes
+  cluster:
+    description:
+    - Name of the cluster to add the host.
+    required: yes
+  folder:
+    description:
+    - Folder name to manage.
+    required: yes
+  hostname:
+    description:
+    - ESXi hostname to manage.
+    required: yes
+  username:
+    description:
+    - ESXi username.
+    required: yes
+  password:
+    description:
+    - ESXi password.
+    required: yes
+  state:
+    description:
+    - Add or remove the folder.
+    default: present
+    choices: [ absent, present ]
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+# Create a folder
+- name: Add a folder to vCenter
+  vmware_folder:
+    hostname: vcsa_host
+    username: vcsa_user
+    password: vcsa_pass
+    datacenter: datacenter
+    cluster: cluster
+    folder: folder
+    state: present
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+instance:
+  description: metadata about the new folder
+  returned: always
+  type: dict
+  sample: None
+'''
+
+try:
+    from pyVmomi import vim, vmodl
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+from ansible.module_utils.vmware import get_all_objs, connect_to_api, vmware_argument_spec, find_datacenter_by_name, \
+    find_cluster_by_name_datacenter, wait_for_task
+from ansible.module_utils.basic import AnsibleModule
+
+
+class VMwareFolder(object):
+    def __init__(self, module):
+        self.module = module
+        self.datacenter = module.params['datacenter']
+        self.cluster = module.params['cluster']
+        self.folder = module.params['folder']
+        self.hostname = module.params['hostname']
+        self.state = module.params['state']
+        self.dc_obj = None
+        self.cluster_obj = None
+        self.host_obj = None
+        self.folder_obj = None
+        self.folder_name = None
+        self.folder_expanded = None
+        self.folder_full_path = []
+        self.content = connect_to_api(module)
+
+    def find_host_by_cluster_datacenter(self):
+        self.dc_obj = find_datacenter_by_name(self.content, self.datacenter)
```

Comment:
Check if self.dc_obj is not None. We have seen "AttributeError" failures in past related to this.